lib/srv/app/tcpserver.go

```diff
@@ -115,3 +119,26 @@ func (s *tcpServer) handleConnection(ctx context.Context, clientConn net.Conn, i
 	}
 	return nil
 }
+
+// ensureZeroTargetPortOrEqualToPortFromURI handles an esoteric edge case where a connection to a
+// single-port TCP app was made with a cert that includes TargetPort meant for multi-port apps.
+//
+// This can happen when the cert was generated before the app spec was changed in a way that
+// transitioned the app from multi-port to single-port. It can also happen due to a programmer error
+// where TargetPort is provided despite the app being single-port.
+func ensureZeroTargetPortOrEqualToPortFromURI(addr *utils.NetAddr, targetPort int) error {
+	if targetPort == 0 {
+		return nil
+	}
+
+	addrPort := addr.Port(0)
+	if addrPort == 0 {
+		return trace.Errorf("missing or invalid port number in URI %q", addr.String())
+	}
+
+	if targetPort != addrPort {
+		return trace.BadParameter("target port is not equal to port number from URI (%d vs %d)", targetPort, addrPort)
```

Comment:
> Maybe the error should say something like "attempt to connect to a TCP app with a multi-port cert where the target port from the cert does not match the port from the app URI (%d vs %q)"?

I prefer that one, but I think we can go a step further and give the user some actionable. Maybe ask them to restart the App Service? Would re-login help here as well?


I'm not a fan of using `x vs y`, because it is not always clear what is `x` and `y`. And you need to check the source code to be sure which is which.
Can we use something like this instead?

```suggestion
		return trace.BadParameter("target port is not equal to port number from URI target_port=%d uri_port=%d", targetPort, addrPort)
```

Reply:
> I prefer that one, but I think we can go a step further and give the user some actionable. Maybe ask them to restart the App Service? Would re-login help here as well?

The problem with giving some actionable here is that this is a client-side issue reported in server-side environment. We have some plans already to catch problems like this one on the client-side too, see https://github.com/gravitational/teleport/pull/46169#discussion_r1742413598 and [the "Incorrect port" section in the RFD](https://github.com/gravitational/teleport/blob/master/rfd/0182-multi-port-tcp-app-access.md#incorrect-port).